### ISSUE TITLE
Allow Body to have a borrowed reader

### ISSUE THREAD

**Issue by Dirbaio** [http-rs/http-types#316]
Thinking about the complexity in async-h1 around cloning the readers, it occurred to me that most of it would go away if Request could have a borrowed body reader.

- Currently async-h1 needs to clone the body reader (wrapping it with `Arc<Mutex<xx>>`) so it can give one clone to the handler, and keep another to discard it if the handler doesn't read it. Giving a borrow to the handler avoids having to clone: once the handler is done the borrow is cleared, so we can discard data from the original reader)
- The body being owned means user code can keep using it even after the response is sent (for example by sending it to another task), which is a logic error (and i'm not entirely sure whether it can corrupt the stream?) If it's borrowed the lifetime prevents that. IMO the reader having a lifetime is "natural": it shouldn't be valid after sending the response.

cc @jbr 

It'd be something like this:

```rust
    pub struct Body<'a> {
        reader: &'a mut dyn AsyncRead + Unpin + Send + Sync,
        ...
    }
    pub struct Request<'a> {
        body: Body<'a>,
        ...
    }
```

**Comment by jbr** [comment]
I agree that this sense in the abstract, but I believe having a lifetime on Request would be a problem for tide, which needs static lifetimes on arguments to endpoints

**Comment by Fishrock123** [comment]
We could maybe allow Body to have an `Arc`'d reader, though. (If that doesn't already work.)

**Comment by Fishrock123** [comment]
@yoshuawuyts I think this is also gated on the whole async closures thing.
